### PR TITLE
 PLANNER-994: Fixed issues related to new Spots & Employee in roster screen

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**
depends on #93

### CODE REVIEW

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/rostergrid/powers/CollisionFreeSubLaneFactory.java [hunk 1]
```diff
@@ -37,7 +37,7 @@
         return blobs.sorted(comparing(Blob::getPositionInGridPixels))
                 .map(this::singletonSubLaneList)
                 .reduce(this::merge)
-                .orElseGet(Collections::emptyList);
+                .orElseGet(() -> new ArrayList<>(Arrays.asList(new SubLane<>())));
```
Comment (tiagobento) [inline optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/rostergrid/powers/CollisionFreeSubLaneFactory.java]: Why did you have to make this list mutable?

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/util/PromiseUtils.java [hunk 2]
```diff
@@ -16,11 +16,60 @@
...
+
+    public <T> Promise<T> promise(PromiseExecutorCallbackFn<T> executor) {
+        return manage(new Promise<>(executor));
+    }
+
+    public <T> Promise<T> manage(Promise<T> promise) {
+        return promise.catch_(this.<T> getDefaultCatch());
+    }
+
+    public <V> CatchOnRejectedCallbackFn<? extends V> getDefaultCatch() {
```
Comment (tiagobento) [inline optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/util/PromiseUtils.java]: I'd make this method private and expose another method called `invokeDefaultErrorHandling` or something with a potentially better name (😝) that does `promiseUtils.getDefaultCatch().onInvoke(i);`. Wdyt?

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/rostergrid/powers/CollisionFreeSubLaneFactory.java (hunk 1 above)
Comment (Christopher-Chianelli) [reply optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/rostergrid/powers/CollisionFreeSubLaneFactory.java]: Needs to be, you are removing elements from it in https://github.com/kiegroup/optashift-employee-rostering/blob/2ebc8b5964a4eef7bb38598d917e4939af3b9d5a/optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/rostergrid/list/ListView.java#L53

File: optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/util/PromiseUtils.java (hunk 2 above)
Comment (Christopher-Chianelli) [reply optashift-employee-rostering-gwtui/src/main/java/org/optaplanner/openshift/employeerostering/gwtui/client/util/PromiseUtils.java]: Good idea, but that is more applicable to #93

**Comment by Christopher-Chianelli** [comment]
@rsynek PromiseUtils is probably impossible to test, since `elemental2.promise.Promise` is native javascript code (meaning the testing framework need to know how to deal with native javascript code). `RosterRestService` does need tests, and so do `EmployeeRestService`, `TenantRestService`, `SpotRestService`, `ShiftRestService` and `RosterGenerator`. I rather do all the REST service tests in a seperate PR (these tests been missing for a while).

**Comment by rsynek** [comment]
@Christopher-Chianelli  Hi! Regarding PromiseUtils - I was afraid so. The REST endpoints could be tested in a separate PR, if you prefer to do so. However, if adding new features has higher priority, this never happens :) So, I am offering a bit different approach:

Each time you touch some area without tests, add some - as scouts leave the camp in a better shape than it was before their arrival.

Following this discipline you don't have to worry about huge PR with tests that usually follows after all features are developed. Remember the Definition of Done.

**Pull request merged**